### PR TITLE
Keycloak unit test

### PR DESCRIPTION
**Pull request by kirtisagarksj**
key cloak unit test

### CODE REVIEW

File: common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        new KeyCloakRsaKeyFetcher()
+            .getPublicKeyFromKeyCloak(
+                KeyCloakConnectionProvider.SSO_URL, KeyCloakConnectionProvider.SSO_REALM);
+    Assert.assertNotNull(key);
+  }
+
+  @Test
+  public void getPublicKeyFromKeyCloakTestFail() throws Exception {
+
+    String jsonString = "InvalidPublicKey";
```
Comment (amolnin) [inline common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakUnitTest.java]: should not fail for invalid public key. failure has to be based on realm.

File: common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  CloseableHttpClient.class,
+  HttpGet.class,
+  CloseableHttpResponse.class,
+  HttpResponse.class,
+  HttpEntity.class,
+  EntityUtils.class,
+  JsonObject.class,
+  JsonParser.class
+})
+public class KeyCloakUnitTest {
```
Comment (bvinayakumar) [inline common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakUnitTest.java]: Rename KeyCloakUnitTest to KeyCloakRsaKeyFetcherTest.

File: common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    PowerMockito.mockStatic(HttpClientBuilder.class);
+    when(HttpClientBuilder.create()).thenReturn(httpClientBuilder);
+    when(httpClientBuilder.build()).thenReturn(client);
+    httpEntity = PowerMockito.mock(HttpEntity.class);
+    PowerMockito.mockStatic(EntityUtils.class);
+    PowerMockito.whenNew(JsonParser.class).withNoArguments().thenReturn(jsonParser);
+  }
+
+  @Test
+  public void getPublicKeyFromKeyCloakTest() throws Exception {
```
Comment (bvinayakumar) [inline common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakUnitTest.java]: Rename all tests in KeyCloakRsaKeyFetcherTest in below format.
e.g. getPublicKeyFromKeyCloakTest -> testGetPublicKeyFromKeyCloakSuccess

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+  Session.class,
+  Metadata.class,
+  CassandraConnectionMngrFactory.class,
+  ResultSet.class,
+  CassandraUtil.class,
+  Cluster.Builder.class,
+  Select.class,
+  Row.class,
+  ColumnDefinitions.class,
+  String.class,
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: Why we need to prepare String class for mocking?

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+  private static PropertiesCache cach = PropertiesCache.getInstance();
+  private static String host = cach.getProperty("contactPoint");
+  private static String port = cach.getProperty("port");
+  private static String cassandraKeySpace = cach.getProperty("keyspace");
+  private static final Cluster.Builder builder = PowerMockito.mock(Cluster.Builder.class);
+  private static BoundStatement boundStatement;
+  private static Select.Selection selectSelection;
+  private static Delete.Selection deleteSelection;
+  private static Delete delete;
+  private static KeyspaceMetadata keyspaceMetadata;
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: is every variable above referred in multiple methods??

Comment (amolnin) [reply cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: For example deleteWhere, builder are required only in initialization method.
Keyspace metadata is used only in specific test-method

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+                ResponseCode.SERVER_ERROR.getResponseCode()));
+
+    List<Row> rows = new ArrayList<>();
+    Row row = Mockito.mock(Row.class);
+    rows.add(row);
+    when(resultSet.all()).thenReturn(rows);
+
+    PowerMockito.whenNew(BoundStatement.class)
+        .withArguments(Mockito.any(PreparedStatement.class))
+        .thenReturn(boundStatement);
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: how is mocking this related, to rest of the mocking rules for the function?

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+    } catch (Exception ex) {
+      exception = ex;
+    }
+    assertTrue(
+        (((ProjectCommonException) exception).getResponseCode())
+            == ResponseCode.SERVER_ERROR.getResponseCode());
+  }
+
+  @Test
+  public void testGetPropertiesValueSuccessById() throws Exception {
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: testGetPropertiesValueByIdSuccess

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+        .withArguments(Mockito.any(PreparedStatement.class))
+        .thenReturn(boundStatement);
+    Response response =
+        operation.getPropertiesValueById(
+            cassandraKeySpace, "address", "123", JsonKey.ID, JsonKey.CITY, JsonKey.ADD_TYPE);
+    assertTrue(response.getResult().size() > 0);
+  }
+
+  @Test
+  public void testGetPropertiesValueFailureById() throws Exception {
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: testGetPropertiesValueByIdFailureWithInvalidProperty

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+
+    Throwable exception = null;
+    PowerMockito.whenNew(BoundStatement.class)
+        .withArguments(Mockito.any(PreparedStatement.class))
+        .thenThrow(
+            new ProjectCommonException(
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: ProjectCommonException.throwServerErrorException - can use - equivalent method throughout the class

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 10]
```diff
@@ -0,0 +1,635 @@
...
+    } catch (Exception ex) {
+      exception = ex;
+    }
+    assertTrue(
+        (((ProjectCommonException) exception).getResponseCode())
+            == ResponseCode.SERVER_ERROR.getResponseCode());
+  }
+
+  @Test
+  public void testGetRecordSuccessById() {
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: Naming convention

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 11]
```diff
@@ -0,0 +1,635 @@
...
+    when(session.execute(where)).thenReturn(resultSet);
+    when(selectBuilder.from(Mockito.anyString(), Mockito.anyString())).thenReturn(selectQuery);
+    when(selectSelection.all()).thenReturn(selectBuilder);
+
+    Response response = operation.getRecordById(cassandraKeySpace, "address", "123");
+    assertTrue(response.getResult().size() > 0);
+  }
+
+  @Test
+  public void testGetRecordFailureById() throws Exception {
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: naming convention

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 12]
```diff
@@ -0,0 +1,635 @@
...
+    } catch (Exception ex) {
+      exception = ex;
+    }
+    assertTrue(
+        (((ProjectCommonException) exception).getResponseCode())
+            == ResponseCode.SERVER_ERROR.getResponseCode());
+  }
+
+  @Test
+  public void testGetRecordSuccessByProperties() throws Exception {
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: Naming convention

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 13]
```diff
@@ -0,0 +1,635 @@
...
+    rows.add(row);
+    when(resultSet.all()).thenReturn(rows);
+
+    Response response =
+        operation.getRecordsByProperty(cassandraKeySpace, "address", JsonKey.ID, list);
+    assertTrue(response.getResult().size() > 0);
+  }
+
+  @Test
+  public void testGetRecordForListFailreByProperties() throws Exception {
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: naming + spelling mistake

File: cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java [hunk 14]
```diff
@@ -0,0 +1,635 @@
...
+  public void testGetRecordFailureByProperties() throws Exception {
+
+    Map<String, Object> map = new HashMap<>();
+    map.put(JsonKey.USER_ID, "USR1");
+    map.put(JsonKey.ADD_TYPE, "addrType");
+
+    List<Row> rows = new ArrayList<>();
+    Row row = Mockito.mock(Row.class);
+    rows.add(row);
+    when(resultSet.all()).thenReturn(rows);
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/CassandraOperationImplTest.java]: Above 4 line snippet repeated in many test-cases - consider maiking it common

File: cassandra-dac/src/test/java/org/sunbird/cassandra/ConnectionManagerTest.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+  Cluster.class,
+  Metadata.class,
+  Cluster.Builder.class,
+  CassandraUtil.class,
+  CassandraConnectionMngrFactory.class,
+})
+@PowerMockIgnore("javax.management.*")
+public class ConnectionManagerTest {
+
+  private static PropertiesCache cach = PropertiesCache.getInstance();
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/ConnectionManagerTest.java]: spelling - cache

File: cassandra-dac/src/test/java/org/sunbird/cassandra/ConnectionManagerTest.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    when(cluster.getMetadata()).thenReturn(metadata);
+    when(Cluster.builder()).thenReturn(builder);
+    when(builder.addContactPoint(Mockito.anyString())).thenReturn(builder);
+    when(builder.withPort(Mockito.anyInt())).thenReturn(builder);
+    when(builder.withProtocolVersion(Mockito.any())).thenReturn(builder);
+    when(builder.withRetryPolicy(Mockito.any())).thenReturn(builder);
+    when(builder.withTimestampGenerator(Mockito.any())).thenReturn(builder);
+    when(builder.withPoolingOptions(Mockito.any())).thenReturn(builder);
+    when(builder.build()).thenReturn(cluster);
+    connectionManager.createConnection(host, port, "cassandra", "password", cassandraKeySpace);
```
Comment (amolnin) [inline cassandra-dac/src/test/java/org/sunbird/cassandra/ConnectionManagerTest.java]: Why we are caling createConnection in before class

File: common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakRsaKeyFetcherTest.java [hunk 17]
```diff
@@ -1,42 +1,96 @@
...
+  CloseableHttpResponse.class,
+  HttpResponse.class,
+  HttpEntity.class,
+  EntityUtils.class,
+  JsonObject.class,
+  JsonParser.class
+})
 public class KeyCloakRsaKeyFetcherTest {
+
   public static final String FAIL_TEST = "fail-check";
```
Comment (amolnin) [inline common-util/src/test/java/org/sunbird/services/sso/impl/KeyCloakRsaKeyFetcherTest.java]: rename to false-realm - variable name as well as value

**Pull request merged**